From 4f1c480c85e3d264d43236cd113bffc6413594d0 Mon Sep 17 00:00:00 2001
From: Xie XiuQi <xiexiuqi@huawei.com>
Date: Wed, 17 Apr 2019 22:01:31 +0800
Subject: [PATCH 2069/2592] [Backport] sched/cputime: add cmdline
 account_irqtime_to_task

hulk inclusion
category: bugfix
bugzilla: 13257

Reference: https://gitee.com/openeuler/kernel/commit/b171c3e6cde7063e53a50b33e04101d25338d87d

-------------------------------

Product team don't want to account irqtime to task.
They want to show the cpu usage to customers as before.

In this patch we add a cmdline option, which is determined by
the product team to turn it on or off.

Signed-off-by: Xie XiuQi <xiexiuqi@huawei.com>
Reviewed-by: Cheng Jian <cj.chengjian@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Zheng Zucheng <zhengzucheng@huawei.com>
---
 kernel/sched/core.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 78bc0d027d22..6ae70fb099b1 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -245,6 +245,15 @@ struct rq *task_rq_lock(struct task_struct *p, struct rq_flags *rf)
  * RQ-clock updating methods:
  */
 
+DEFINE_STATIC_KEY_FALSE(account_irqtime_to_task);
+static int __init setup_account_irqtime(char *str)
+{
+	static_branch_enable(&account_irqtime_to_task);
+
+	return 0;
+}
+__setup("account-irqtime-to-task", setup_account_irqtime);
+
 static void update_rq_clock_task(struct rq *rq, s64 delta)
 {
 /*
@@ -254,6 +263,9 @@ static void update_rq_clock_task(struct rq *rq, s64 delta)
 	s64 __maybe_unused steal = 0, irq_delta = 0;
 
 #ifdef CONFIG_IRQ_TIME_ACCOUNTING
+	if (static_branch_unlikely(&account_irqtime_to_task))
+		goto out;
+
 	irq_delta = irq_time_read(cpu_of(rq)) - rq->prev_irq_time;
 
 	/*
@@ -276,6 +288,8 @@ static void update_rq_clock_task(struct rq *rq, s64 delta)
 
 	rq->prev_irq_time += irq_delta;
 	delta -= irq_delta;
+
+out:
 #endif
 #ifdef CONFIG_PARAVIRT_TIME_ACCOUNTING
 	if (static_key_false((&paravirt_steal_rq_enabled))) {
-- 
2.17.1

